File: 101_200/113_Path_Sum_II/Solution.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public List<List<Integer>> pathSum(TreeNode root, int sum) {
        List<List<Integer>> ans = new ArrayList();
        List<Integer> cur = new ArrayList();
        dfs(root, sum, cur, ans);
        return ans;
    }
    
    public void dfs(TreeNode root, Integer target, List<Integer> curr, List<List<Integer>> ans) {
        if (root == null) return;
        curr.add(root.val);
        if (root.left == null && root.right == null && root.val == target) {
            ans.add(new ArrayList(curr));
        }
        if (root.left != null) dfs(root.left, target-root.val, curr, ans);
        if (root.right != null) dfs(root.right, target -root.val, curr, ans);
        curr.remove(curr.size() - 1);
    }
}

File: 101_200/113_Path_Sum_II/Solution.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def pathSum(self, root: TreeNode, sum: int) -> List[List[int]]:
        ans = []
        t = []

        def dfs(n, target, cur):
            if not n:
                return
            cur.append(n.val)
            if not n.left and not n.right and target == n.val:
                ans.append(cur.copy())
            if n.left:
                dfs(n.left, target - n.val, cur)
            if n.right:
                dfs(n.right, target - n.val, cur)
            cur.pop()
        dfs(root, sum, t)
        return ans
